Fix 27338 - Allow passing an actor system to the typed.ActorTestKit

Reklund3 opened the pull request:
Ran into two Binary compatability issue with this pull request locally, they were:
```
[error] akka-actor-testkit-typed: Failed binary compatibility check against com.typesafe.akka:akka-actor-testkit-typed_2.12:2.6.2! Found 2 potential problems
[error]  * method config()com.typesafe.config.Config in class akka.actor.testkit.typed.scaladsl.ActorTestKit does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("akka.actor.testkit.typed.scaladsl.ActorTestKit.config")
[error]  * method this(java.lang.String,com.typesafe.config.Config,scala.Option)Unit in class akka.actor.testkit.typed.scaladsl.ActorTestKit's type is different in current version, where it is (java.lang.String,akka.actor.setup.ActorSystemSetup,scala.Option)Unit instead of (java.lang.String,com.typesafe.config.Config,scala.Option)Unit
[error]    filter with: ProblemFilters.exclude[IncompatibleMethTypeProblem]("akka.actor.testkit.typed.scaladsl.ActorTestKit.this")
```
Adding two apply methods to accept an actor system or a name and Actor System Setup.
Additionally, I updated the ActorTestKit class to take an ActorSystemSetup instead of the config and changed the `internalSystem` variable to construct an ActorSystem using the apply constructor that accepts a `ActorSystemSeup`.
Updated the remaining apply methods to convert the passed config into an ActorSystemSetup to pass into the updated class.
Updated the test specs to remove the config variable from validation as the config variable no longer exists.

References #27338
https://discuss.lightbend.com/t/typed-actor-testkit/4569

lightbend-cla-validator commented:
At least one pull request committer is not linked to a user. See https://help.github.com/en/articles/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user

lightbend-cla-validator commented:
At least one pull request committer is not linked to a user. See https://help.github.com/en/articles/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user

Reklund3 commented:
I have signed the CLA.

patriknw commented on akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/javadsl/ActorTestKit.scala:
```diff
@@ -35,6 +36,36 @@ object ActorTestKit {
+  /**
+   * Create a testkit from the provided actor system.
+   *
+   * It will create an [[akka.actor.typed.ActorSystem]] with the provided actor name,
```
this factory doesn't have a name parameter

patriknw commented on akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/scaladsl/ActorTestKit.scala:
```diff
@@ -70,7 +109,7 @@ object ActorTestKit {
     new ActorTestKit(
       name = TestKitUtils.testNameFromCallStack(classOf[ActorTestKit]),
-      config = customConfig,
+      ActorSystemSetup.create(BootstrapSetup(customConfig)),
```
could use `apply` to be more Scala style
```
ActorSystemSetup(BootstrapSetup(customConfig))
```

patriknw commented on akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/scaladsl/ActorTestKit.scala:
```diff
@@ -40,8 +43,41 @@ object ActorTestKit {
+   * Config loaded from `system.settings.config` if that exists, otherwise
+   * using default configuration from the reference.conf resources that ship with the Akka libraries.
+   */
+  def apply(system: ActorSystem[_]): ActorTestKit = {
```
This is only using the name of that `system`, which I don't see the purpose of.
We also want to add support for passing in a custom ActorSystem and use that instead of creating an ActorSystem. That means that the testkit still needs to spawn ActorTestKitGuardian in that system.
So when passing in a system it should
```
val testKitGuardian: ActorRef[ActorTestKitGuardian.TestKitCommand] =
  system.systemActorOf(ActorTestKitGuardian.testKitGuardian, name = "test")
```

The nice thing is that for the other option when the testkit is creating the ActorSystem that is also an `ActorRef[ActorTestKitGuardian.TestKitCommand]`

```
val testKitGuardian: ActorRef[ActorTestKitGuardian.TestKitCommand] =
  ActorSystem(ActorTestKitGuardian.testKitGuardian, name, config)
```

patriknw commented on akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/scaladsl/ActorTestKit.scala:
```diff
@@ -40,8 +43,41 @@ object ActorTestKit {
+   * Config loaded from `application-test.conf` if that exists, otherwise
+   * using default configuration from the reference.conf resources that ship with the Akka libraries.
+   */
+  def apply(name: String, actorSystemSetup: ActorSystemSetup): ActorTestKit = {
```
If we have the support for passing in an ActorSystem as described above there might be less need of having this ActorSystemSetup variant since the user has full control of how the ActorSystem is created, including its Setup.

lightbend-cla-validator commented:
At least one pull request committer is not linked to a user. See https://help.github.com/en/articles/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user

lightbend-cla-validator commented:
At least one pull request committer is not linked to a user. See https://help.github.com/en/articles/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user

lightbend-cla-validator commented:
At least one pull request committer is not linked to a user. See https://help.github.com/en/articles/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user

lightbend-cla-validator commented:
At least one pull request committer is not linked to a user. See https://help.github.com/en/articles/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user

lightbend-cla-validator commented:
At least one pull request committer is not linked to a user. See https://help.github.com/en/articles/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user

lightbend-cla-validator commented:
At least one pull request committer is not linked to a user. See https://help.github.com/en/articles/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user

Reklund3 committed:
Add apply methods for typed ActorTestKit allowing for passing an ActorSystem or ActorSystemSetup (#27338).

Added an apply method that takes a akka.actor.typed.ActorSystem and generates a ActorTestKit.
Added an apply method that takes a string and akka.actor.setup.ActorSystemSetup and generates an ActorTestKit

Modified the internal ActorTestKit class to use an ActorSystemSetup instead of a config and updated the other apply methods to generate an ActorSystemSetup in place of the config.
This was an attempt to maintain compatability.

added the related test specs for the new apply methods and updated the apply methods that changed.

Reklund3 committed:
Add apply methods for typed ActorTestKit Javadsl (#27338) fix author

Reklund3 committed:
Correcting apply method doc comments for Javadsl apply methods (#27338) fix author

Reklund3 committed:
running scala format for (27338) fix author

Reklund3 committed:
Addressing review comments for (fix-27338)

Reklund3 committed:
Updating the apply methods to construct the actor system and pass it to the ActorTestKit. (27338) fix author

Reklund3 committed:
running scala format (27338) fix author

Reklund3 committed:
Cleaning up the doc comment to be more reflective of the actual method. (27338) fix author

Reklund3 committed:
Updating the javadsl doc comment (27338) fix author

Reklund3 committed:
Updating the path for the ClusterReceptionistSpec for changes to the ActorTestKit (fix-27338)

chbatey commented on akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/javadsl/ActorTestKit.scala:
```diff
@@ -35,6 +35,21 @@ object ActorTestKit {
+   * Create a testkit from the provided actor system.
+   *
+   * It will use the provided [[akka.actor.typed.ActorSystem]] and create a ActorTestKit,
+   * e.g. threads will include the name of the provided actor system.
```
this comment line isn't relevant for this create method but I see most are incorrect in this file so we can follow up with a fix for that

chbatey commented on akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/scaladsl/ActorTestKit.scala:
```diff
@@ -131,19 +155,19 @@ object ActorTestKit {
-  @InternalApi private[akka] val internalSystem: ActorSystem[ActorTestKitGuardian.TestKitCommand] =
-    ActorSystem(ActorTestKitGuardian.testKitGuardian, name, config)
+  @InternalApi private[akka] val internalTestKitGuardian: ActorRef[ActorTestKitGuardian.TestKitCommand] =
+    internalSystem.systemActorOf(ActorTestKitGuardian.testKitGuardian, "test")
```
This isn't used as a "Guardian" any more. It is always spawned as a system actor so we should rename this. We can do it in this PR or we can follow up later.

chbatey commented on akka-actor-testkit-typed/src/test/scala/akka/actor/testkit/typed/scaladsl/ActorTestKitSpec.scala:
```diff
@@ -21,6 +23,16 @@ class ActorTestKitSpec extends ScalaTestWithActorTestKit with AnyWordSpecLike wi
       system.name should ===("ActorTestKitSpec")
     }
 
+    "generate a default name & ActorSystemSetup from the provided actor" in {
```
i think this test verifies that the provided actor system is used rather than anything related to ActorSystemSetup, rename?

patriknw commented on akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/scaladsl/ActorTestKit.scala:
```diff
@@ -39,10 +39,26 @@ object ActorTestKit {
+   * Config loaded from `system.settings.config` if that exists, otherwise
+   * using default configuration from the reference.conf resources that ship with the Akka libraries.
+   */
+  def apply(system: ActorSystem[_]): ActorTestKit =
```
we could also add corresponding to `ScalaTestWithActorTestKit` and `TestKitJunitResource` so an ActorSystem can be passed in also when using those

patriknw commented on akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/scaladsl/ActorTestKit.scala:
```diff
@@ -56,7 +72,9 @@ object ActorTestKit {
   def apply(name: String): ActorTestKit =
-    new ActorTestKit(name = TestKitUtils.scrubActorSystemName(name), config = ApplicationTestConfig, settings = None)
+    new ActorTestKit(
+      ActorSystem(ActorTestKitGuardian.testKitGuardian, TestKitUtils.scrubActorSystemName(name), ApplicationTestConfig),
```
In this case the `ActorTestKitGuardian.testKitGuardian` isn't used, since it's always spawning another `internalTestKitGuardian`. It would be possible to get around that by passing in `testKitGuardian` in the constructor of ActorTestKit.

```
final class ActorTestKit private[akka] (val internalSystem: ActorSystem[_], internalTestKitGuardian: ActorRef[ActorTestKitGuardian.TestKitCommand], settings: Option[TestKitSettings]) {
```

then the two apply cases would look like:
```
  def apply(name: String): ActorTestKit = {
    val system = ActorSystem(ActorTestKitGuardian.testKitGuardian, TestKitUtils.scrubActorSystemName(name), ApplicationTestConfig)
    new ActorTestKit(
      system,
      system,
      settings = None)
  }

  def apply(system: ActorSystem[_]): ActorTestKit = {
    val testKitGuardian = system.systemActorOf(ActorTestKitGuardian.testKitGuardian, "test")
    new ActorTestKit(system, testKitGuardian, settings = None)
  }
```

patriknw reviewed (commented):
looking good, a suggestion of how to spawn the ActorTestKitGuardian...

Reklund3 committed:
Addressing review comments for fix-27338

Creating the Guardian in the apply and passing it to the ActorTestKit
Attempt to cleanup doc comments for the updated apply methods
Adding the apply method that takes an actor system to ScalaTestWithActorTestkit and TestKitJunitResource

Reklund3 committed:
reverting changes to ClusterReceptionistSpec I made that were incorrect.

Reklund3 committed:
Updating apply doc comments and ActortestKitSpec test description for passing and actor system.

patriknw commented on akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/scaladsl/ActorTestKit.scala:
```diff
@@ -131,20 +163,20 @@ object ActorTestKit {
+final class ActorTestKit private[akka] (
+    val internalSystem: ActorSystem[_],
+    internalTestKitGuardian: ActorRef[ActorTestKitGuardian.TestKitCommand],
+    settings: Option[TestKitSettings]) {
```
`val name` and `val config` are now gone. For compatibility reasons you can add
```
  val name = internalSystem.name
  val config = internalSystem.settings.config
```

I think that should also get rid of the two of the binary compatibility warnings.

The other warnings can be filtered out by adding the following to new file `akka-actor-testkit-typed/src/main/mima-filters/2.6.4.backwards.excludes/27338-ActorTestKit-system.excludes`

```
# #27338 allow passing ActorSystem to ActorTestKit
ProblemFilters.exclude[IncompatibleSignatureProblem]("akka.actor.testkit.typed.scaladsl.ActorTestKit.internalSystem")
ProblemFilters.exclude[IncompatibleMethTypeProblem]("akka.actor.testkit.typed.scaladsl.ActorTestKit.this")
```

patriknw reviewed (commented):
good, just something around the binary compatibility remaining...

Reklund3 committed:
Adding the name and config variables back to the ActorTestKit class and the two mima lines for binary compatibility.

patriknw reviewed (approved):
LGTM, thanks

Reklund3 commented:
Thank you.

Reklund3 commented:
When I ran mima locally I still received two potential warnings for Akka-actor-typed:2.6.2

patriknw commented:
ok, I'll check

patriknw commented:
Those have been fixed in master. PR validation runs on a merge of master so should be all good. We'll see when it has completed.
